core/src/main/scala/kafka/raft/KafkaNetworkChannel.scala

```diff
@@ -44,7 +44,10 @@ object KafkaNetworkChannel {
       case fetchRequest: FetchRequestData =>
         // Since we already have the request, we go through a simplified builder
         new AbstractRequest.Builder[FetchRequest](ApiKeys.FETCH) {
-          override def build(version: Short): FetchRequest = new FetchRequest(fetchRequest, version)
+          override def build(version: Short): FetchRequest = {
+            FetchRequest.maybeDownGradeReplicaState(fetchRequest, version)
+            new FetchRequest(fetchRequest, version)
+          }
```

Comment:
As we have logic in here now, I wonder if defining a `SimpleBuilder` in the `FetchRequest` class would make sense. This would allow use to write unit test for it as well. What do you think?

Reply:
Ack.